xds/src/main/java/io/grpc/xds/XdsNameResolver.java

```diff
@@ -517,18 +503,16 @@ public void run() {
       }
     }
 
-    private long generateHash(List<HashPolicy> hashPolicies, Map<String, String> headers) {
+    private long generateHash(List<HashPolicy> hashPolicies, Metadata headers) {
       Long hash = null;
       for (HashPolicy policy : hashPolicies) {
         Long newHash = null;
         if (policy.type() == HashPolicy.Type.HEADER) {
-          if (headers.containsKey(policy.headerName())) {
-            String value = headers.get(policy.headerName());
-            if (policy.regEx() != null && policy.regExSubstitution() != null) {
-              value = policy.regEx().matcher(value).replaceAll(policy.regExSubstitution());
-            }
-            newHash = hashFunc.hashAsciiString(value);
+          String value = getHeaderValue(headers, policy.headerName());
+          if (value != null && policy.regEx() != null && policy.regExSubstitution() != null) {
+            value = policy.regEx().matcher(value).replaceAll(policy.regExSubstitution());
           }
+          newHash = hashFunc.hashAsciiString(value);
```

Comment:
This changes the hashing behavior when value == null. Is that expected?

Reply:
I think the fix is broken too, as now there must be a regex. Before only the header presence was required. This probably deserves a test.

Reply:
My bad, messed up the case for not having header substitution rule. Fixed and added a test.